File: src/ir_printer.ml
open Ir
open Printf

let verbose = false
let if_verbose s = if verbose then s else ""

let rec string_of_val_type = function
  | Int(w) -> sprintf "i%d" w
  | UInt(w) -> sprintf "u%d" w
  | Float(w) -> sprintf "f%d" w
  | IntVector(w, n) -> sprintf "i%dx%d" w n
  | UIntVector(w, n) -> sprintf "u%dx%d" w n
  | FloatVector(w, n) -> sprintf "f%dx%d" w n

and string_of_op = function
  | Add -> "+"
  | Sub -> "-"
  | Mul -> "*"
  | Div -> "/"
  | Mod -> "%"
  | Min -> "(min)"
  | Max -> "(max)"

and string_of_cmp = function
  | EQ -> "=="
  | NE -> "!="
  | LT -> "<"
  | LE -> "<="
  | GT -> ">"
  | GE -> ">="

and can_skip_parens op1 op2 = match (op1, op2) with
  | (Add, Mul) -> true
  | (Add, Div) -> true
  | (Sub, Mul) -> true
  | (Sub, Div) -> true
  | (Mul, Div) -> true
  | (Mul, Mul) -> true
  | (Add, Add) -> true
  | (Add, Sub) -> true
  | _ -> false

and string_of_expr = function
  | IntImm(i) -> string_of_int i
  | UIntImm(i) -> string_of_int i ^ "u"
  | FloatImm(f) -> string_of_float f
  | Cast(t, e) -> string_of_val_type t ^ "(" ^ string_of_expr e ^")"
  | Bop(outer_op, Bop(inner_op, ll, lr), r) when can_skip_parens outer_op inner_op ->
      "(" ^ string_of_expr ll ^ string_of_op inner_op ^ string_of_expr lr ^ string_of_op outer_op ^ string_of_expr r ^ ")"
  | Bop(outer_op, l, Bop(inner_op, rl, rr)) when can_skip_parens outer_op inner_op ->
      "(" ^ string_of_expr l ^ string_of_op outer_op ^ string_of_expr rl ^ string_of_op inner_op ^ string_of_expr rr ^ ")"
  | Bop(Add, l, IntImm x) when x < 0 -> string_of_expr (l -~ (IntImm (-x)))
  | Bop(op, l, r) -> "(" ^ string_of_expr l ^ string_of_op op ^ string_of_expr r ^ ")"
  | Cmp(op, l, r) -> "(" ^ string_of_expr l ^ string_of_cmp op ^ string_of_expr r ^ ")"
  | Select(c, t, f) ->
    "(" ^ string_of_expr c ^ "?" ^ string_of_expr t ^ ":" ^ string_of_expr f ^ ")"
  | Var(t, s) -> s ^ if_verbose ("<" ^ string_of_val_type t ^ ">")
  | Load(t, b, i) -> string_of_buffer b ^ "[" ^ string_of_expr i ^ "]"
  | Call(ct, rt, name, args) -> name ^ if_verbose ("<" ^ string_of_val_type rt ^ ">") ^ "(" ^ (String.concat ", " (List.map string_of_expr args)) ^ ")"
  | MakeVector l -> "vec[" ^ (String.concat ", " (List.map string_of_expr l)) ^ "]"
  | Broadcast(e, n) -> "broadcast[" ^ string_of_expr e ^ ", " ^ string_of_int n ^ "]"
  | Ramp(b, s, n) -> "ramp[" ^ string_of_expr b ^ ", " ^ string_of_expr s ^ ", " ^ string_of_int n ^ "]"
  | ExtractElement(a, b) -> "(" ^ string_of_expr a ^ "@" ^ string_of_expr b ^ ")"
  | Let(n, a, b) -> "(let " ^ n ^ " = " ^ string_of_expr a ^ " in " ^ string_of_expr b ^ ")"
  (* | Debug(e, prefix, args) -> "Debug(" ^ string_of_expr e ^ ", " ^ prefix ^ " " ^  
      (String.concat ", " (List.map string_of_expr args)) ^ ")" *)
  | Debug (e, _, _) -> string_of_expr e
  | Or (a, b) -> "(" ^ string_of_expr a ^ " or " ^ string_of_expr b ^ ")"
  | And (a, b) -> "(" ^ string_of_expr a ^ " and " ^ string_of_expr b ^ ")"
  | Not (a) -> "(not " ^ string_of_expr a ^ ")"
  (* | _ -> "<<UNHANDLED>>" *)

and string_of_stmt stmt = 
  let rec string_stmt p stmt =
    let sp = p ^ "  " in
    match stmt with
      (* | If(e, s) -> "if (" ^ string_of_expr e ^ ") " ^ string_of_stmt s *)
      (* | IfElse(e, ts, fs) -> "if (" ^ string_of_expr e ^ ") " ^ string_of_stmt ts ^ 
       " else " ^ string_of_stmt fs *)
      | For(name, min, n, order, s) -> 
          (p ^ (sprintf "%s (%s: %s, %s) {\n" 
                  (if order then "for" else "pfor")
                  name (string_of_expr min) (string_of_expr n)) ^
             (string_stmt sp s) ^ 
             p ^ "}\n")
      | Block(stmts) -> 
          (p ^ "{" ^ "\n" ^
             String.concat "" (List.map (string_stmt sp) stmts) ^
             p ^ "}\n")

      | Store(e, b, i) -> 
          (p ^ string_of_buffer b ^ "[" ^ string_of_expr i ^ "] = \n" ^
             sp ^ string_of_expr e ^ ";\n")             
      | Provide (e, f, args) ->
          (p ^ f ^ "(" ^ (String.concat ", " (List.map string_of_expr args)) ^ ") = \n" ^
             sp ^ string_of_expr e ^ ";\n")
      | LetStmt (name, value, stmt) ->
          (p ^ "let " ^ name ^ " = " ^ string_of_expr value ^ "\n" ^ 
             string_stmt p stmt)            
      | Pipeline(name, ty, size, produce, consume) -> 
          (p ^ "produce " ^ name ^ "[" ^ string_of_expr size ^ "] {\n" ^ 
             string_stmt sp produce ^ 
             p ^ "}\n" ^
             string_stmt p consume)
      | Print (m, l) -> p ^ "Print(" ^ m ^ (String.concat ", " (List.map string_of_expr l)) ^ ")\n"
      | Assert (e, str) -> p ^ "Assert(" ^ (string_of_expr e) ^ ", " ^ str ^ ")\n"
          
  in
  string_stmt "" stmt
    
(*
and string_of_reduce_op = function
  | AddEq -> "+="
  | SubEq -> "-="
  | MulEq -> "*="
  | DivEq -> "/="
 *)

and string_of_buffer b = b

and string_of_toplevel (n, a, s) = n ^ "(" ^ String.concat ", " (List.map string_of_arg a) ^ ") =\n" ^ (string_of_stmt s)

and string_of_arg = function 
  | Buffer (b) -> b
  | Scalar (s, t) -> (string_of_val_type t) ^ " " ^ s

and string_of_definition (name, args, rtype, body) =
  let string_of_arglist f args = "(" ^ String.concat ", " (List.map f args) ^ ")" in
  let s = name ^
          (string_of_arglist (fun (_,n) -> n) args) ^
          " -> " ^ (string_of_val_type rtype)
  in
  match body with
    | Pure(e) -> s ^ " = " ^ (string_of_expr e)

    (* Reduction consists of initializer, update location, update function, and iteration domain for the update *)
    | Reduce (init, update_idx, update_func, reduction_domain) ->
        (* First write the init rule *)
        let init_s = string_of_definition(name, args, rtype, Pure(init)) in
        (* Then write the update rule *)
        let update_s = name ^ ".reduce" ^
                       (string_of_arglist (fun idx -> (string_of_expr idx)) update_idx) ^
                       " <- " ^
                       update_func ^ (string_of_arglist (fun (name,_,_) -> name) reduction_domain) ^ "\n" ^
                       "\tover " ^
                       (string_of_arglist
                         (fun (name, min, size) ->
                           name ^ " : [" ^ (string_of_expr min) ^ ".." ^ (string_of_expr size) ^ "]")
                         reduction_domain)
                       in
        init_s ^ "\n" ^ update_s

and string_of_environment env =
  let (_,defs) = List.split (Environment.bindings env) in
  String.concat "\n" (List.map string_of_definition defs)

File: cpp/IRPrinter.cpp
#include "IRPrinter.h"
#include "IROperator.h"
#include "IR.h"

#include <iostream>

namespace HalideInternal {

    using std::endl;

    void IRPrinter::test() {
        Type i32 = Int(32);
        Type f32 = Float(32);
        Expr x = new Var(i32, "x");
        Expr y = new Var(i32, "y");
        std::cout << (x + 3) * (y / 2 + 17) << endl;

        Stmt store = new Store("buf", (x * 17) / (x - 3), y - 1);
        Stmt for_loop = new For("x", -2, y + 2, For::Parallel, store);
        vector<Expr> args(1); args[0] = x % 3;
        Expr call = new Call(i32, "buf", args, Call::Halide);
        Stmt store2 = new Store("out", call + 1, x);
        Stmt for_loop2 = new For("x", 0, y, For::Vectorized , store2);
        Stmt pipeline = new Pipeline("buf", for_loop, Stmt(), for_loop2);
        Stmt assertion = new AssertStmt(y > 3, "y is greater than 3");
        Stmt block = new Block(assertion, pipeline);
        Stmt let_stmt = new LetStmt("y", 17, block);
        Stmt allocate = new Allocate("buf", f32, 1023, let_stmt);

        std::cout << allocate << endl;
    }

    ostream &operator<<(ostream &out, Type type) {
        switch (type.t) {
        case Type::Int:
            out << 'i';
            break;
        case Type::UInt:
            out << 'u';
            break;
        case Type::Float:
            out << 'f';
            break;
        default:
            assert(false && "Malformed type");
        }
        out << type.bits;
        if (type.width > 1) out << 'x' << type.width;
        return out;
    }

    ostream &operator<<(ostream &out, For::ForType type) {
        switch (type) {
        case For::Serial:
            out << "for";
            break;
        case For::Parallel:
            out << "parallel";
            break;
        case For::Unrolled:
            out << "unrolled";
            break;
        case For::Vectorized:
            out << "vectorized";
            break;
        default:
            assert(false && "Malformed for type");
        }
        return out;
    }

    ostream &operator<<(ostream &stream, Expr ir) {
        IRPrinter p(ir, stream);
        return stream;
    }

    ostream &operator<<(ostream &stream, Stmt ir) {
        IRPrinter p(ir, stream);
        return stream;
    }

    void do_indent(ostream &out, int indent) {
        for (int i = 0; i < indent; i++) out << ' ';
    }

    IRPrinter::IRPrinter(Expr ir, ostream &s) : stream(s), indent(0) {
        ir.accept(this);
    }

    IRPrinter::IRPrinter(Stmt ir, ostream &s) : stream(s), indent(0) {
        ir.accept(this);
    }
    
    void IRPrinter::visit(const IntImm *op) {
        stream << op->value;
    }
    
    void IRPrinter::visit(const FloatImm *op) {
        stream << op->value;
    }
    
    void IRPrinter::visit(const Cast *op) { 
        stream << op->type << '(';
        op->value.accept(this);
        stream << ')';
    }
    
    void IRPrinter::visit(const Var *op) {
        // omit the type
        stream << op->name;
    }
    
    void IRPrinter::visit(const Add *op) {
        stream << '(';
        op->a.accept(this);
        stream << " + ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const Sub *op) {
        stream << '(';
        op->a.accept(this);
        stream << " - ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const Mul *op) {
        stream << '(';
        op->a.accept(this);
        stream << "*";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const Div *op) {
        stream << '(';
        op->a.accept(this);
        stream << "/";
        op->b.accept(this);
        stream << ')';
    }
            
    void IRPrinter::visit(const Mod *op) {
        stream << '(';
        op->a.accept(this);
        stream << " % ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const Min *op) {
        stream << "min(";
        op->a.accept(this);
        stream << ", ";
        op->b.accept(this);
        stream << ")";
    }

    void IRPrinter::visit(const Max *op) {
        stream << "max(";
        op->a.accept(this);
        stream << ", ";
        op->b.accept(this);
        stream << ")";
    }

    void IRPrinter::visit(const EQ *op) {
        stream << '(';
        op->a.accept(this);
        stream << " == ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const NE *op) {
        stream << '(';
        op->a.accept(this);
        stream << " != ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const LT *op) {
        stream << '(';
        op->a.accept(this);
        stream << " < ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const LE *op) {
        stream << '(';
        op->a.accept(this);
        stream << " <= ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const GT *op) {
        stream << '(';
        op->a.accept(this);
        stream << " > ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const GE *op) {
        stream << '(';
        op->a.accept(this);
        stream << " >= ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const And *op) {
        stream << '(';
        op->a.accept(this);
        stream << " && ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const Or *op) {
        stream << '(';
        op->a.accept(this);
        stream << " || ";
        op->b.accept(this);
        stream << ')';
    }

    void IRPrinter::visit(const Not *op) {
        stream << '!';
        op->a.accept(this);
    }

    void IRPrinter::visit(const Select *op) {
        stream << "select(";
        op->condition.accept(this);
        stream << ", ";
        op->true_value.accept(this);
        stream << ", ";
        op->false_value.accept(this);
        stream << ")";
    }

    void IRPrinter::visit(const Load *op) {
        stream << op->buffer << "[";
        op->index.accept(this);
        stream << "]";
    }

    void IRPrinter::visit(const Ramp *op) {
        stream << "ramp(";
        op->base.accept(this);
        stream << ", ";
        op->stride.accept(this);
        stream << ", " << op->width << ")";
    }

    void IRPrinter::visit(const Broadcast *op) {
        stream << "broadcast(";
        op->value.accept(this);
        stream << ", " << op->width << ")";
    }

    void IRPrinter::visit(const Call *op) {
        stream << op->buffer << "(";
        for (size_t i = 0; i < op->args.size(); i++) {
            op->args[i].accept(this);
            if (i < op->args.size() - 1) {
                stream << ", ";
            }
        }
        stream << ")";
    }

    void IRPrinter::visit(const Let *op) {
        stream << "(let " << op->name << " = ";
        op->value.accept(this);
        stream << " in ";
        op->body.accept(this);
        stream << ")";
    }

    void IRPrinter::visit(const LetStmt *op) {
        do_indent(stream, indent);
        stream << "let " << op->name << " = ";
        op->value.accept(this);
        stream << endl;

        do_indent(stream, indent);
        op->body.accept(this);
    }

    void IRPrinter::visit(const PrintStmt *op) {
        do_indent(stream, indent);
        stream << "print(" << op->prefix;
        for (size_t i = 0; i < op->args.size(); i++) {
            stream << ", ";
            op->args[i].accept(this);
        }
        stream << ")" << endl;
    }

    void IRPrinter::visit(const AssertStmt *op) {
        do_indent(stream, indent);
        stream << "assert(";
        op->condition.accept(this);
        stream << ", \"" << op->message << "\")" << endl;
    }

    void IRPrinter::visit(const Pipeline *op) {
        indent += 2;

        do_indent(stream, indent - 2);
        stream << "produce " << op->buffer << " {" << endl;
        op->produce.accept(this);

        if (op->update.defined()) {
            do_indent(stream, indent - 2);
            stream << "} update {" << endl;
            op->update.accept(this);            
        }
        
        do_indent(stream, indent - 2);
        stream << "} consume {" << endl;
        op->consume.accept(this);
        
        do_indent(stream, indent - 2);        
        stream << "}" << endl;

        indent -= 2;
    }

    void IRPrinter::visit(const For *op) {

        do_indent(stream, indent);
        stream << op->for_type << " (" << op->name << ", ";
        op->min.accept(this);
        stream << ", ";
        op->extent.accept(this);
        stream << ") {" << endl;
        
        indent += 2;
        op->body.accept(this);
        indent -= 2;

        do_indent(stream, indent);
        stream << "}" << endl;
    }

    void IRPrinter::visit(const Store *op) {
        do_indent(stream, indent);
        stream << op->buffer << "[";
        op->index.accept(this);
        stream << "] = ";
        op->value.accept(this);
        stream << endl;
    }

    void IRPrinter::visit(const Provide *op) {
        do_indent(stream, indent);
        stream << op->buffer << "(";
        for (size_t i = 0; i < op->args.size(); i++) {
            op->args[i].accept(this);
            if (i < op->args.size() - 1) stream << ", ";
        }
        stream << ") = ";
        op->value.accept(this);
        stream << endl;
    }

    void IRPrinter::visit(const Allocate *op) {
        do_indent(stream, indent);
        stream << "allocate " << op->buffer << "[" << op->type << " * ";
        op->size.accept(this);
        stream << "]" << endl;
        op->body.accept(this);

        do_indent(stream, indent);
        stream << "free " << op->buffer << endl;
    }

    void IRPrinter::visit(const Realize *op) {
        do_indent(stream, indent);
        stream << "realize " << op->buffer << "(";
        for (size_t i = 0; i < op->bounds.size(); i++) {
            stream << "[";
            op->bounds[i].first.accept(this);
            stream << ", ";
            op->bounds[i].second.accept(this);
            stream << "]";
            if (i < op->bounds.size() - 1) stream << ", ";
        }
        stream << ") {" << endl;

        indent += 2;
        op->body.accept(this);
        indent -= 2;

        do_indent(stream, indent);
        stream << "}" << endl;
    }

    void IRPrinter::visit(const Block *op) {
        op->first.accept(this);
        if (op->rest.defined()) op->rest.accept(this);
    }

    
}
